stl/inc/deque

```diff
@@ -986,19 +986,58 @@ public:
     }
 
     void shrink_to_fit() {
-        size_type _Oldcapacity = _Block_size * _Mapsize();
-        size_type _Newcapacity = _Oldcapacity / 2;
+        if (empty()) {
+            if (_Map() != nullptr) {
+                _Reset_map();
+            }
+            return;
+        }
+
+        const auto _First_block_ptr = _Map() + static_cast<_Map_difference_type>(_Myoff() / _Block_size);
+        const auto _Last_block_ptr =
+            _Map() + static_cast<_Map_difference_type>((_Myoff() + _Mysize() - 1) / _Block_size + 1);
+        const auto _Used_block_count = static_cast<size_type>(_Last_block_ptr - _First_block_ptr);
 
-        if (_Newcapacity < _Block_size * _Minimum_map_size) {
-            _Newcapacity = _Block_size * _Minimum_map_size;
+        // deallocate unused blocks
+        for (auto _Block_it = _Map(); _Block_it != _First_block_ptr; ++_Block_it) {
+            if (*_Block_it != nullptr) {
+                _Getal().deallocate(*_Block_it, _Block_size);
+                *_Block_it = nullptr;
+            }
+        }
+        for (auto _Block_it = _Last_block_ptr; _Block_it != _Map() + _Map_distance(); ++_Block_it) {
+            if (*_Block_it != nullptr) {
+                _Getal().deallocate(*_Block_it, _Block_size);
+                *_Block_it = nullptr;
+            }
         }
 
-        if ((empty() && _Mapsize() > 0)
-            || (!empty() && size() <= _Newcapacity && _Newcapacity < _Oldcapacity)) { // worth shrinking, do it
-            deque _Tmp(
-                _STD make_move_iterator(_Unchecked_begin()), _STD make_move_iterator(_Unchecked_end()), _Getal());
-            swap(_Tmp);
+        size_type _New_block_count = 1; // should be power of 2
+        for (; _New_block_count < _Used_block_count || _New_block_count < _Minimum_map_size; _New_block_count *= 2)
+            ;
+        if (_New_block_count >= _Mapsize()) {
+            return;
         }
+
+        // worth shrinking the internal map, do it
+
+        _Alpty _Almap(_Getal());
+        const auto _New_map = _Almap.allocate(_New_block_count);
+
+        _Orphan_all(); // the map can be shrinked, invalidate all iterators
+
+        const auto _Copy_end = _STD _Uninitialized_copy_unchecked(
+            _First_block_ptr, _Last_block_ptr, _New_map); // transfer the ownership of blocks to pointers in the new map
+        _STD _Uninitialized_value_construct_n_unchecked1(
+            _Copy_end, _New_block_count - _Used_block_count); // zero out the rest of the new map
+        _STD fill(
+            _First_block_ptr, _Last_block_ptr, nullptr); // zero out the old map because the ownership is transferred
+
+        _Reset_map();

```

Comment:
```
_STD fill(
    _First_block_ptr, _Last_block_ptr, nullptr); // zero out the old map because the ownership is transferred

_Reset_map();
```

This can be just deallocating `_Map()`: `_Almap.deallocate(_Map(), _Mapsize());`